README.md



# future.BatchJobs: A Future API for Parallel and Distributed Processing using BatchJobs

![Life cycle: superseded](vignettes/imgs/lifecycle-superseded-blue.svg)

_NOTE: The [BatchJobs](https://cran.r-project.org/package=BatchJobs) package is deprecated in favor of the [batchtools](https://cran.r-project.org/package=batchtools) package. Because of this, it is recommended to use the [future.batchtools](https://cran.r-project.org/package=future.batchtools) package instead of this package. This [future.BatchJobs](https://cran.r-project.org/package=future.BatchJobs) package is formally deprecated and will be archive on CRAN._


## Introduction

The [future] package provides a generic API for using futures in R.
A future is a simple yet powerful mechanism to evaluate an R expression
and retrieve its value at some point in time.  Futures can be resolved
in many different ways depending on which strategy is used.
There are various types of synchronous and asynchronous futures to
choose from in the [future] package.

This package, [future.BatchJobs], provides a type of futures that
utilizes the [BatchJobs] package.  This means that _any_ type of
backend that the BatchJobs package supports can be used as a future.
More specifically, future.BatchJobs will allow you or users of your
package to leverage the compute power of high-performance computing
(HPC) clusters via a simple switch in settings - without having to
change any code at all.

For instance, if BatchJobs is properly configures, the below two
expressions for futures `x` and `y` will be processed on two different
compute nodes:
```r
> library("future.BatchJobs")
> plan(batchjobs_custom)
>
> x %<-% { Sys.sleep(5); 3.14 }
> y %<-% { Sys.sleep(5); 2.71 }
> x + y
[1] 5.85
```
This is obviously a toy example to illustrate what futures look like
and how to work with them.

A more realistic example comes from the field of cancer research
where very large data FASTQ files, which hold a large number of short
DNA sequence reads, are produced.  The first step toward a biological
interpretation of these data is to align the reads in each sample
(one FASTQ file) toward the human genome.  In order to speed this up,
we can have each file be processed by a separate compute node and each
node we can use 24 parallel processes such that each process aligns a
separate chromosome.  Here is an outline of how this nested parallelism
could be implemented using futures.
```r
library("future")
library("listenv")
## The first level of futures should be submitted to the
## cluster using BatchJobs.  The second level of futures
## should be using multiprocessing, where the number of
## parallel processes is automatically decided based on
## what the cluster grants to each compute node.
plan(list(batchjobs_custom, multisession))

## Find all samples (one FASTQ file per sample)
fqs <- dir(pattern="[.]fastq$")

## The aligned results are stored in BAM files
bams <- listenv()

## For all samples (FASTQ files) ...
for (ss in seq_along(fqs)) {
  fq <- fqs[ss]

  ## ... use futures to align them ...
  bams[[ss]] %<-% {
    bams_ss <- listenv()
	## ... and for each FASTQ file use a second layer
	## of futures to align the individual chromosomes
    for (cc in 1:24) {
      bams_ss[[cc]] %<-% htseq::align(fq, chr=cc)
    }
	## Resolve the "chromosome" futures and return as a list
    as.list(bams_ss)
  }
}
## Resolve the "sample" futures and return as a list
bams <- as.list(bams)
```
Note that a user who do not have access to a cluster could use the same script processing samples sequentially and chromosomes in parallel on a single machine using:
```r
plan(list(sequential, multisession))
```
or samples in parallel and chromosomes sequentially using:
```r
plan(list(multisession, sequential))
```

For an introduction as well as full details on how to use futures,
please consult the package vignettes of the [future] package.



## Choosing BatchJobs backend
The future.BatchJobs package implements a generic future wrapper
for all BatchJobs backends.  Below are the most common types of
BatchJobs backends.


| Backend                 | Description                                                              | Alternative in future package
|:------------------------|:-------------------------------------------------------------------------|:------------------------------------
| _generic:_              |                                                                          |
| `batchjobs_custom`      | Uses custom BatchJobs configuration script files, e.g. `.BatchJobs.R`    | N/A
| _predefined:_           |                                                                          |
| `batchjobs_torque`      | Futures are evaluated via a [TORQUE] / PBS job scheduler                 | N/A
| `batchjobs_slurm`       | Futures are evaluated via a [Slurm] job scheduler                        | N/A
| `batchjobs_sge`         | Futures are evaluated via a [Sun/Oracle Grid Engine (SGE)] job scheduler | N/A
| `batchjobs_lsf`         | Futures are evaluated via a [Load Sharing Facility (LSF)] job scheduler  | N/A
| `batchjobs_openlava`    | Futures are evaluated via an [OpenLava] job scheduler                    | N/A
| `batchjobs_interactive` | synchronous evaluation in the calling R environment                      | `plan(transparent)`
| `batchjobs_local`       | synchronous evaluation in a separate R process (on current machine)      | `plan(cluster, workers="localhost")`

In addition to the above, there is also `batchjobs_multicore` (which on Windows and Solaris falls back to `batchjobs_local`), which runs BatchJobs tasks asynchronously in background R sessions (sic!) on the current machine.  We _advise to not use_ this and instead use `multisession` of the [future] package.  For details, see `help("batchjobs_multicore")`.


### Examples

Below are two examples illustrating how to use `batchjobs_custom` and `batchjobs_torque` to configure the BatchJobs backend.  For further details and examples on how to configure BatchJobs, see the [BatchJobs configuration] wiki page.

### Example: A .BatchJobs.R file using local BatchJobs
The most general way of configuring BatchJobs is via a `.BatchJobs.R` file.
This file should be located in the current directory or in the user's
home directory.  For example, as an alternative to `batchjobs_local`,
we can manually configure local BatchJobs futures a `.BatchJobs.R` file
that contains
```r
cluster.functions <- makeClusterFunctionsLocal()
```
This will then be found and used when specifying
```r
> plan(batchjobs_custom)
```
To specify this BatchJobs configuration file explicitly, one can use
```r
> plan(batchjobs_custom, pathname="./.BatchJobs.R")
```

This follows the naming convention set up by the BatchJobs package.



### Example: A .BatchJobs.*.tmpl template file for TORQUE / PBS
To configure BatchJobs for job schedulers we need to setup a template
file that is used to generate the script used by the scheduler.
This is what a template file for TORQUE / PBS may look like:
```sh
## Job name:
#PBS -N <%= job.name %>

## Merge standard error and output:
#PBS -j oe

## Resource parameters:
<% for (name in names(resources)) { %>
#PBS -l <%= name %>=<%= resources[[name]] %>
<% } %>

## Run R:
R CMD BATCH --no-save --no-restore "<%= rscript %>" /dev/stdout
```
If this template is saved to file `.BatchJobs.torque.tmpl` in the
working directory or the user's home directory, then it will be
automatically located and loaded when doing:
```r
> plan(batchjobs_torque)
```
Resource parameters can be specified via argument `resources` which should be a named list and is passed as is to the template file.  For example, to request that each job would get allotted 12 cores (one a single machine) and up to  5 GiB of memory, use:
```r
> plan(batchjobs_torque, resources=list(nodes="1:ppn=12", vmem="5gb"))
```

To specify the `resources` argument at the same time as using nested future strategies, one can use `tweak()` to tweak the default arguments.  For instance,
```r
plan(list(
  tweak(batchjobs_torque, resources=list(nodes="1:ppn=12", vmem="5gb")),
  multisession
))
```
causes the first level of futures to be submitted via the TORQUE job scheduler requesting 12 cores and 5 GiB of memory per job.  The second level of futures will be evaluated using multiprocessing using the 12 cores given to each job by the scheduler.

A similar filename format is used for the other types of job schedulers supported.  For instance, for Slurm the template file should be named `.BatchJobs.slurm.tmpl` in order for
```r
> plan(batchjobs_slurm)
```
to locate the file automatically.  To specify this template file explicitly, use argument `pathname`, e.g.
```r
> plan(batchjobs_slurm, pathname="./.BatchJobs.slurm.tmpl")
```


Note that it is still possible to use a `.BatchJobs.R` and load the template file using a standard BatchJobs approach for maximum control.  For further details and examples on how to configure BatchJobs per se, see the [BatchJobs configuration] wiki page.



## Demos
The [future] package provides a demo using futures for calculating a
set of Mandelbrot planes.  The demo does not assume anything about
what type of futures are used.
_The user has full control of how futures are evaluated_.
For instance, to use local BatchJobs futures, run the demo as:
```r
library("future.BatchJobs")
plan(batchjobs_local)
demo("mandelbrot", package="future", ask=FALSE)
```


[BatchJobs]: https://cran.r-project.org/package=BatchJobs
[brew]: https://cran.r-project.org/package=brew
[future]: https://cran.r-project.org/package=future
[future.BatchJobs]: https://cran.r-project.org/package=future.BatchJobs
[BatchJobs configuration]: https://github.com/tudo-r/BatchJobs/wiki/Configuration
[TORQUE]: https://en.wikipedia.org/wiki/TORQUE
[Slurm]: https://en.wikipedia.org/wiki/Slurm_Workload_Manager
[Sun/Oracle Grid Engine (SGE)]: https://en.wikipedia.org/wiki/Oracle_Grid_Engine
[Load Sharing Facility (LSF)]: https://en.wikipedia.org/wiki/Platform_LSF
[OpenLava]: https://en.wikipedia.org/wiki/OpenLava


## Installation
R package future.BatchJobs is available on [CRAN](https://cran.r-project.org/package=future.BatchJobs) and can be installed in R as:
```r
install.packages("future.BatchJobs")
```


### Pre-release version

To install the pre-release version that is available in Git branch `develop` on GitHub, use:
```r
remotes::install_github("HenrikBengtsson/future.BatchJobs", ref="develop")
```
This will install the package from source.  

## Contributions

This Git repository uses the [Git Flow](https://nvie.com/posts/a-successful-git-branching-model/) branching model (the [`git flow`](https://github.com/petervanderdoes/gitflow-avh) extension is useful for this).  The [`develop`](https://github.com/HenrikBengtsson/future.BatchJobs/tree/develop) branch contains the latest contributions and other code that will appear in the next release, and the [`master`](https://github.com/HenrikBengtsson/future.BatchJobs) branch contains the code of the latest release, which is exactly what is currently on [CRAN](https://cran.r-project.org/package=future.BatchJobs).

Contributing to this package is easy.  Just send a [pull request](https://help.github.com/articles/using-pull-requests/).  When you send your PR, make sure `develop` is the destination branch on the [future.BatchJobs repository](https://github.com/HenrikBengtsson/future.BatchJobs).  Your PR should pass `R CMD check --as-cran`, which will also be checked by <a href="https://travis-ci.org/HenrikBengtsson/future.BatchJobs">Travis CI</a> and <a href="https://ci.appveyor.com/project/HenrikBengtsson/future-batchjobs">AppVeyor CI</a> when the PR is submitted.

We abide to the [Code of Conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/) of Contributor Covenant.


## Software status

| Resource      | CRAN        | GitHub Actions      | Travis CI       | AppVeyor CI      |
| ------------- | ------------------- | ------------------- | --------------- | ---------------- |
| _Platforms:_  | _Multiple_          | _Multiple_          | _Linux & macOS_ | _Windows_        |
| R CMD check   | <a href="https://cran.r-project.org/web/checks/check_results_future.BatchJobs.html"><img border="0" src="http://www.r-pkg.org/badges/version/future.BatchJobs" alt="CRAN version"></a> |        | <a href="https://travis-ci.org/HenrikBengtsson/future.BatchJobs"><img src="https://travis-ci.org/HenrikBengtsson/future.BatchJobs.svg" alt="Build status"></a>   | <a href="https://ci.appveyor.com/project/HenrikBengtsson/future-batchjobs"><img src="https://ci.appveyor.com/api/projects/status/github/HenrikBengtsson/future.BatchJobs?svg=true" alt="Build status"></a> |
| Test coverage |                     |                     | <a href="https://codecov.io/gh/HenrikBengtsson/future.BatchJobs"><img src="https://codecov.io/gh/HenrikBengtsson/future.BatchJobs/branch/develop/graph/badge.svg" alt="Coverage Status"/></a>     |                  |


NEWS
Package: future.BatchJobs
=========================

Version: 0.17.0-9000 [2021-01-08]

SIGNIFICANT CHANGES:

 * The 'future.BatchJobs' package has been archived on CRAN as of 2021-01-08.
 

Version: 0.17.0 [2021-01-02]

SIGNIFICANT CHANGES:

 * The 'future.BatchJobs' package is deprecated and will soon be archived on
   CRAN in favor of the 'future.batchtools' package.

 * Lazy batchjobs futures only creates the internal BatchJobs registry when
   the future is launched.
 

Version: 0.16.2 [2019-09-27]

NEW FEATURES:

 * Debug messages are now prepended with a timestamp.
 
BUG FIXES:

 * Argument 'workers' could not be a function.

 * Argument 'workers' of type character was silently accepted and effectively
   interpreted as workers = length(workers).


Version: 0.16.1 [2019-01-03]

FIXES:

 * Made internal code agile to upcoming changes in the future package on
   how a captured error is represented.

BUG FIXES:

 * resolve() on a lazy batchtools future would stall and never return.


Version: 0.16.0 [2018-07-09]

NEW FEATURES:

 * The batchjobs_* backends support the handling of the standard output
   as implemented in future (>= 1.9.0).

FIXES:

 * Since the BatchJobs package is no longer maintained it is has not been
   updated according to the new API of the DBI package. This caused several
   warnings to be produced when using BatchJobs futures.  These warnings
   are now suppressed (to avoid all the noise from these warnings).

SOFTWARE QUALITY:

 * Removed internal workaround for BatchJobs (< 1.7).

CODE REFACTORING:

 * Preparing for futures to gather a richer set of results from BatchJobs
   backends.

DEPRECATED AND DEFUNCT:

 * Removed defunct and hidden argument 'args' to BatchJobsFuture.


Version: 0.15.0 [2017-09-10]

NEW FEATURES:

 * The error message for expired BatchJobs futures now include the last few
   lines of the logged output, which sometimes includes clues on why the future
   expired.  For instance, if a TORQUE / PBS job use more than the allocated
   amount of memory it might be terminated by the scheduler leaving the message
   "PBS: job killed: vmem 1234000 exceeded limit 1048576" in the output.

 * print() for BatchtoolsFuture returns the object invisibly.

BUG FIXES:

 * Calling future_lapply() with functions containing globals part of non-default
   packages would when using batchtools futures give an error complaining that
   the global is missing. This was due to updates in future (>= 1.4.0) that
   broke this package.

 * loggedOutput() for BatchtoolsFuture would always return NULL unless an error
   had occurred.

DEPRECATED AND DEFUNCT:

 * Removed defunct batchjobs() and backend() functions.

 * Passing hidden argument 'args' to BatchJobsFuture is now defunct.
 

Version: 0.14.1 [2017-05-30]

DOCUMENTATION:

 * Removing remaining references to 'eager' (now using 'sequential').

SOFTWARE QUALITY:

 * Testing future_lapply() for batchjobs backends.

 * TESTS: No longer testing with (deprecated) 'lazy' or 'eager' backends.
 

Version: 0.14.0 [2017-03-18]

NEW FEATURES:

 * The number of jobs one can add to the queues of HPC schedulers is in
   principle unlimited, which is why the number of available workers for
   such batchjobs_* backends is reported as +Inf.  However, as the number
   of workers is used by future_lapply() to decide how many futures should
   be used to best partition the elements, this means that future_lapply()
   will always use one future per element.  Because of this, it is now
   possible to specify plan(batchjobs_*, workers = n) where 'n' is the
   target number of workers.
 
 * Option 'future.wait.timeout' (replaces 'future.wait.times') specifies
   the maximum waiting time for BatchJobs futures to finish before
   generating a timeout error.

BUG FIX:

 * Futures with globals would give an error when using the developers
   version of BatchJobs (> 1.6) with globals (<= 0.7.1).  Package now
   requires future (>= 1.4.0) which in turn requires globals (>= 0.8.0).

DEPRECATED AND DEFUNCT:

 * Previously deprecated batchjobs() and backend() functions are defunct.
   Instead, use one of the corresponding batchjobs_*() functions.
 

Version: 0.13.1 [2016-10-20]

NEW FEATURES:

 * Added argument 'job.delay' to batchjobs_*() futures for passing
   it as is to BatchJobs::submitJobs() used when launching futures.
   
 * Added argument 'label' to batchjobs_*() futures which are reflected
   in the job name listed by schedulers.  Because of limitation is
   BatchJobs, not all characters in the labels are supported and are
   therefore dropped in the job names.

GLOBAL VARIABLES:

 * GLOBALS: Now globals can be specified explicitly.


Version: 0.13.0 [2016-08-02]

NEW FEATURES:

 * Added argument 'resources' to batchjobs_*() functions for passing
   it to the BatchJobs template (as variable 'resources').
   
 * ROBUSTNESS: value() now launches the future iff not already done.
   Added protection from launching a future more than once.


Version: 0.12.1 [2016-06-26]

DOCUMENTATION:

 * Advising against _multicore BatchJobs_ futures, because there
   is a risk for long waiting times due to starvation.
   This is a limitation of the BatchJobs package.

BUG FIX:

 * Multicore BatchJobs futures are not supported on Solaris Unix and
   now falls back to local BatchJobs futures (as on Windows).
   This is a limitation of the BatchJobs package.


Version: 0.12.0 [2016-06-25]

NEW FEATURES:

 * Added predefined batchjobs_local(), batchjobs_interactive(),
   batchjobs_multicore(), batchjobs_lsf(), batchjobs_openlava(),
   batchjobs_sge(), batchjobs_slurm(), batchjobs_torque() and
   batchjobs_custom() futures.
   
 * Added nbrOfWorkers() for BatchJobs futures.

 * CLEANUP: Now "Loading required package: BatchJobs [...]", which is
   outputted when the first BatchJobs future is created, is suppressed.

DEPRECATED AND DEFUNCT:

 * Removed non-used completed(), failed() and expired() for BatchJobs objects.


R/backend.R
#' Switch backend to be used for asynchronous processing
#'
#' \emph{This function is deprecated.}
#'
#' @param what
#'   A \code{character} \code{vector} of preferred backends to be used.
#'   See Section Details below for supported backends.
#'   If \code{NULL} (default), the currently default backend is returned.
#'   If \code{"reset"}, the backend is reset to \code{"default"}.
#'   If \code{"default"}, the default backend according to alias
#'      \code{"default"} is used (see Details).
#'   If \code{"aliases"}, all registered aliases are returned.
#' @param \ldots Named character arguments specifying custom aliases
#'            for character sets of backends.
#' @param quietly If TRUE, messages are suppressed.
#'
#' @return Returns the name of the backend used, or a list of named aliases.
#'
#' @details
#' \itemize{
#'  \item \code{".BatchJobs.R"} -
#'    fully configurable batch processing based on a {.BatchJobs.R}
#'    configuration file (searched for in the current directory and
#'    if not found there then in the home directory).
#'    Alternatively, one can specify a specific pathname, e.g.
#'    \code{backend("~/shared/.BatchJobs.R")}.#
#'
#'  \item \code{"interactive"} -
#'    non-parallel processing in the \emph{current} R session (but still
#'    with all calls effectively evaluated as within \code{local()}).
#'    \emph{This backend is always supported.}
#'
#'  \item \code{"local"} -
#'    non-parallel processing in a separate R process.
#'    \emph{This backend is always supported.}
#'    \emph{This is the fallback used if none of the requested
#'          backends are supported.}
#'
#'  \item \code{"multicore"} -
#'    parallel processing using all available cores on the local machine.
#'    The number of available cores is inferred using
#'    \code{\link[future]{availableCores}()}.
#'    Note: Multicore processing is not supported on Windows.
#'    If explicitly specified, then an informative warning will be given,
#'    and it will be ignored.
#'
#'  \item \code{"multicore=<n>"} -
#'    as \code{"multicore"}, but uses (at most) \code{<n>} cores, e.g.
#'    \code{backend("multicore=4")}.
#'  \item \code{"multicore-<d>"} -
#'    as \code{"multicore"}, but uses \code{<d>} less cores, e.g.
#'    \code{backend("multicore-2")} to use all but two cores
#'    (and always at least one).
#' }
#'
#' In addition specify an individual backend, e.g. \code{backend("multicore")},
#' it is also possible to specify a set of possible backends to use, e.g.
#' \code{backend(c(".BatchJobs.R", "multicore"))}.
#' The first supported backend will be used.
#' If none are supported, then the \code{"local"} backend will be used.
#'
#' It is possible to define a custom set of backends using "aliases".
#' For instance, \code{backend(spare=c("multicore-2", "local"))}
#' \emph{defines} the \code{"spare"} backend \emph{set}.  To use this set,
#' do \code{backend("spare")}.  One predefined alias set exists at startup,
#' i.e. \code{backend(default = c(".BatchJobs.R", "multicore-1",
#'            "multicore", "local"))}.
#'
#' @export
#' @importFrom future availableCores
#' @importFrom tools file_path_as_absolute
#' @importFrom utils file_test
#' @importFrom BatchJobs makeClusterFunctionsMulticore makeClusterFunctionsLocal makeClusterFunctionsInteractive
backend <- local({
  aliases = list(
    default = c(".BatchJobs.R", "multicore-1", "local")
  )
  last = NULL

  ## WORKAROUND: See comment below.
  dyld_envs <- list()

  function(what=NULL, ..., quietly=TRUE) {
    ## Import private functions from BatchJobs
    sourceConfFiles <- importBatchJobs("sourceConfFiles")
    assignConf <- importBatchJobs("assignConf")
    readConfs <- importBatchJobs("readConfs")

    debug <- getOption("future.debug", FALSE)

    ## Set custom aliases?
    custom <- list(...)
    if (length(custom) > 0L) {
      names <- names(custom)
      if (is.null(names)) {
        stop("Trying to call backend() without named arguments.")
      }
      for (name in names) {
        value <- custom[[name]]
        if (!is.null(value) && !is.character(value)) {
          stop("Backend aliases must be character vectors: ", mode(value))
        }
        aliases[[name]] <<- value
      }
      return(invisible(aliases))
    }

    if (is.null(what)) {
      return(last)
    } else if (identical(what, "aliases")) {
      return(aliases)
    } else if (identical(what, "reset")) {
      what <- "default"
    }

    explicit_what <- !missing(what)
    dropped <- NULL

    ## Expand aliases
    if (length(aliases) > 0L) {
      names <- names(aliases)
      what <- lapply(what, FUN=function(what) {
        if (what %in% names) aliases[[what]] else what
      })
      what <- unlist(what, use.names=FALSE)
    }

    ## BatchJobs configuration?
    if (length(what) >= 1L) {
      ## A "global" or a specific config file?
      ## NOTE: For a file in the current directory, use ./.BatchJobs.R
      if (what[1L] == ".BatchJobs.R") {
        if (!hasUserClusterFunctions()) {
          dropped <- c(dropped, what[1L])
          what <- what[-1L]
        }
      } else if (file_test("-f", what[1L])) {
        pathname <- file_path_as_absolute(what[1L])
        if (!hasUserClusterFunctions(pathname)) {
          stop("The specified file does not specify BatchJobs cluster functions: ", what[1L])
        }
      }
    }


    ## Check if multicore processing is supported or makes sense
    ncpus0 <- availableCores()
    for (kk in seq_along(what)) {
      if (!grepl("^multicore", what[kk])) next

      ## Multicore processing is not supported on Windows
      if (.Platform$OS.type == "windows") dropped <- c(dropped, what[kk])

      ## Default number of cores to use
      ncpus <- ncpus0

      if (grepl("^multicore=", what[kk])) {
        ncpus <- suppressWarnings(as.integer(gsub("^multicore=", "", what[kk])))
        if (!is.finite(ncpus) || ncpus < 1L) {
          stop("Invalid number of cores specified: ", sQuote(what[kk]))
        }
      } else {
        ## Leave some cores for other things?
        if (grepl("^multicore-", what[kk])) {
          save <- suppressWarnings(as.integer(gsub("^multicore-", "", what[kk])))
          if (!is.finite(save) || save < 0L) {
            stop("Invalid number of cores specified: ", sQuote(what[kk]))
          }
          ncpus <- min(1L, ncpus-save)  ## At least one core
        }
      }

      ## If only one core is available or one one was requested, then
      ## there is no point using multicore processing.
      ## NOTE: This also solves the problem that BatchJobs will wait
      ## endlessly if 'ncpus=1' and there are already 3 R processes
      ## running on the same machine, which happens for instance when
      ## 'R CMD check' runs tests.  For more details on this "feature",
      ## see runBatchJobs:::getWorkerSchedulerStatus(). /HB 2016-05-16
      if (ncpus < 2L) {
        dropped <- c(dropped, what[kk])
      }
    } ## for (kk ...)

    what <- setdiff(what, dropped)

    ## Always fall back to using the 'local' configuration
    if (length(what) == 0L) what <- "local"

    ## The final choice
    what <- what[1L]


    if (debug) mprintf("backend(): what='%s'\n", what)

    ## Inform about dropped requests?
    if (length(dropped) > 0L && explicit_what && getOption("future.on_unkown_backend", "ignore") == "warn") {
      warning(sprintf("Some of the preferred backends (%s) are either not available, pointless (e.g. multicore=1), or not supported on your operating system ('%s'). Will use the following backend instead: %s", paste(sQuote(dropped), collapse=", "), .Platform$OS.type, sQuote(what)))
    }

    ## Load specific or global BatchJobs config file?
    if (file_test("-f", what)) {
      .Deprecated(new=sprintf("plan(batchjobs_conf, pathnames='%s')", what))
      if (debug) mprintf("backend(): file='%s'\n", what)
      conf <- sourceConfFiles(what)
      if (debug) {
        mprintf("Setting BatchJobs configuration:\n")
        mstr(as.list(conf))
      }
      assignConf(conf)
      ## Record last used
      last <<- what
      return(what)
    } else if (what == ".BatchJobs.R") {
      .Deprecated(new=sprintf("plan(batchjobs_conf)", what))
      if (debug) mprintf("backend(): First available '.BatchJobs.R'\n")
      if (quietly) {
        suppressPackageStartupMessages(readConfs())
      } else {
        readConfs()
      }
      ## Record last used
      last <<- what
      return(what)
    }

    if (debug) mprintf("backend(): Finding action for what='%s'\n", what)

    cluster.functions <- NULL
    if (grepl("^multicore", what)) {
      ## Sanity check (see above)
      stopifnot(ncpus0 >= 2L)

      if (grepl("^multicore=", what)) {
        ncpus <- suppressWarnings(as.integer(gsub("^multicore=", "", what)))
        if (!is.finite(ncpus) || ncpus < 1L) {
          stop("Invalid number of cores specified: ", sQuote(what))
        }
        .Deprecated(new=sprintf("plan(batchjobs_multicore, workers=%d)", ncpus))

        if (ncpus > ncpus0) {
          warning(sprintf("The number of specific cores (%d) is greater than (%d) what is available according to availableCores(). Will still try to use this requested backend: %s", ncpus, ncpus0, sQuote(what)))
        }
      } else {
        ncpus <- ncpus0

        ## Leave some cores for other things?
        if (grepl("^multicore-", what)) {
          save <- suppressWarnings(as.integer(gsub("^multicore-", "", what)))
          .Deprecated(new=sprintf("plan(batchjobs_multicore, workers=availableCores()-%d)", ncpus))
          ncpus <- ncpus - save
        } else {
          .Deprecated(new="plan(batchjobs_multicore)")
        }
      }

      ## Sanity check (see above)
      stopifnot(ncpus >= 2L)


      ## WORKAROUND:
      ## On some OS X systems, a system call to 'ps' may output an error message
      ## "dyld: DYLD_ environment variables being ignored because main executable
      ##  (/bin/ps) is setuid or setgid" to standard error that is picked up by
      ## BatchJobs which incorrectly tries to parse it.  By unsetting all DYLD_*
      ## environment variables, we avoid this message.  For more info, see:
      ## * https://github.com/tudo-r/BatchJobs/issues/117
      ## * https://github.com/HenrikBengtsson/future.BatchJobs/issues/59
      ## /HB 2016-05-07
      dyld_envs <<- tryCatch({
        envs <- list()
        res <- system2("ps", stdout=TRUE, stderr=TRUE)
        if (any(grepl("DYLD_", res))) {
          envs <- Sys.getenv()
          envs <- envs[grepl("^DYLD_", names(envs))]
          if (length(envs) > 0L) lapply(names(envs), FUN=Sys.unsetenv)
        }
        envs
      }, error = function(ex) list())


      ## PROBLEM 1:
      ## BatchJobs' multicore cluster functions tries to be responsive to the overall
      ## CPU load of the machine (as reported by Linux command 'uptime') and it will
      ## not submit new jobs if the load is greater than its 'max.load' parameter.
      ## This parameter is by default set to one less than number of available cores
      ## on the machine (as reported by parallel::detectCores()).  This way it tries
      ## to leave some leeway for other processes avoiding clogging up the machine.
      ## If 'mc.cores' is set, that it taken as the number of available cores instead.
      ## See BatchJobs:::makeWorker() for code.  However, the CPU load is still relative
      ## to the true number of cores available.  In other words, the check that the
      ## observed CPU load is less than 'max.load' (== mc.cores-1) is not correct and
      ## may end up never to be meet, resulting in an endless waiting to submit jobs.
      ## WORKAROUND:
      ## A better estimate may be to set 'max.load' to be parallel::detectCores()-1.
      ## However, it appears that that may also stall the processing in some cases.
      ## Because of this, we set the limit to +Inf.  This should be alright because
      ## max.jobs=ncpus (also the default if not specified).  If a user wish to use
      ## other settings, this can be done via a custom .BatchJobs.R file.
      ## /HB 2016-05-16
      max.load <- +Inf
      if (debug) mprintf("backend(): makeClusterFunctionsMulticore(ncpus=%d, max.jobs=%d, max.load=%g)\n", ncpus, ncpus, max.load)
      cluster.functions <- makeClusterFunctionsMulticore(ncpus=ncpus, max.jobs=ncpus, max.load=max.load)
    } else if (what == "local") {
      .Deprecated(new="plan(batchjobs_local)")
      cluster.functions <- makeClusterFunctionsLocal()
    } else if (what == "interactive") {
      .Deprecated(new="plan(batchjobs_interactive)")
      cluster.functions <- makeClusterFunctionsInteractive()
    } else {
      stop("Unknown backend: ", sQuote(what))
    }


    ## WORKAROUND: Undo above multicore OS X workaround, iff ever done.
    if (!grepl("^multicore", what) && length(dyld_envs) > 0L) {
      do.call(Sys.setenv, args=as.list(dyld_envs))
    }

    conf <- makeBatchJobsConf(cluster.functions)

    ## Use it!
    if (debug) {
      mprintf("Setting BatchJobs configuration:\n")
      mstr(as.list(conf))
    }
    assignConf(conf)

    ## Record last used
    last <<- what

    what
  }
}) # backend()


# Check whether user specifies 'cluster.functions' in one of
# the .BatchJobs.R files that BatchJobs loads.
hasUserClusterFunctions <- function(pathnames=NULL, debug=FALSE) {
  ns <- getNamespace("BatchJobs")
  findConfigs <- get("findConfigs", mode="function", envir=ns)
  sourceConfFiles <- get("sourceConfFiles", mode="function", envir=ns)

  if (is.null(pathnames)) {
    pathnames <- findConfigs()
    ## Drop ... ? Where does that file exists? /HB 2015-05-19
    pathnames <- pathnames[basename(pathnames) != "BatchJobs_global_config.R"]
  }
  if (debug) mprint(pathnames)

  suppressPackageStartupMessages({
    config <- sourceConfFiles(pathnames)
  })

  exists("cluster.functions", mode="list", envir=config)
}


makeBatchJobsConf <- function(cluster.functions, ...) {
  getBatchJobsConf <- importBatchJobs("getBatchJobsConf")

  conf <- getBatchJobsConf()

  conf$cluster.functions <- cluster.functions
  conf$mail.start <- "none"
  conf$mail.done <- "none"
  conf$mail.error <- "none"
  conf$db.driver <- "SQLite"
  conf$db.options <- list()
  conf$default.resources <- list()
  conf$debug <- FALSE
  conf$raise.warnings <- FALSE
  conf$staged.queries <- TRUE
  conf$max.concurrent.jobs <- Inf
  conf$fs.timeout <- NA_real_

  ## Sanity check
  stopifnot(is.environment(conf))

  conf
} ## makeBatchJobsConf()
